cupy/cuda/cub.pyx

```diff
@@ -244,6 +245,11 @@ cdef _cub_reduce_dtype_compatible(x_dtype, int op, dtype=None,
         support_dtype = CUB_support_dtype
     else:
         return False
+
+    if int(device.get_compute_capability()) >= 53:
+        if x_dtype == numpy.float16:
+            return True
```

Comment:
For 1, I think your suggestion is faster than the current implementation due to the reduction of device API calls. However, I think we should check the CC based on the current device, thus I added CC detecting on calling time. Or should we create the all `CUB_support_dtype` and `CUB_sum_support_dtype` on each device basis? And get the corresponding list with `_get_device_id` without calling Device API. What do you think?

For 2, I think so too. I'll fix my PR based on your comment. Thank you.

Reply:
> However, I think we should check the CC based on the current device, thus I added CC detecting on calling time. 

Ah, I see your concern now. This is a bit trickier than I thought, hmm...

If I am right, you are worried about the possibility that perhaps at the import time the device 0 has a higher CC (>=53) than the actual device with CC < 53 to be used for reductions later?

> Or should we create the all `CUB_support_dtype` and `CUB_sum_support_dtype` on each device basis? And get the corresponding list with `_get_device_id` without calling Device API. What do you think?

Yes. I think this would be a sensible approach. I would then cache the generated `CUB_support_dtype` and `CUB_sum_support_dtype` lists based on device ids. So, they become dicts (key = id int or string, value = list/tuple without or with float16), and when checking CUB dtypes we only call `get_compute_capability()` if the key lookup via `get()` returns `None`. 

Something like this is perhaps OK?
```python
# at module level
CUB_support_dtype_without_half = (...)
CUB_support_dtype_with_half = CUB_support_dtype_without_half + (numpy.float16,)
CUB_support_dtype = {}

CUB_sum_support_dtype_without_half = (...)
CUB_sum_support_dtype_with_half = CUB_sum_support_dtype_without_half + (numpy.float16,)
CUB_sum_support_dtype = {}

# lookup using device_id in CUB helpers
dtypes = CUB_support_dtype.get(device_id)
if dtypes is None:
    if int(device.get_compute_capability()) >= 53:
        dtypes = CUB_support_dtype_with_half
    else:
        dtypes = CUB_support_dtype_without_half
    CUB_support_dtype[device_id] = dtypes
# ...and test with the dtypes tuple

# same applies to CUB_sum_support_dtype
```
What do you think?

Reply:
> If I am right, you are worried about the possibility that perhaps at the import time the device 0 has a higher CC (>=53) than the actual device with CC < 53 to be used for reductions later?

Yes. There are two devices on a node but they may have different CC.

> So, they become dicts (key = id int or string, value = list/tuple without or with float16), and when checking CUB dtypes we only call get_compute_capability() if the key lookup via get() returns None.

It's a good idea. I'll fix my PR based on this idea. Thank you!

Reply:
I found that `device.get_device_id()` still calls `cudaGetDevice` so we cannot reduce the CUDA Device API invocation. And also compute capabilities are cached in `device.pyx` so calling this isn't time-consuming. IMHO, this dictionary-based implementation is a little complex but no performance gain. What do you think? Do you have any idea to reduce `device.get_device_id()` calls?

Reply:
My idea:
- check all devices cc
- if all cc is same, we skip calling `device.get_device_id()`
- if not (probably rare), we call `device.get_device_id()` on each cub call.

Reply:
> I found that `device.get_device_id()` still calls `cudaGetDevice` so we cannot reduce the CUDA Device API invocation. 

Ouch!

> And also compute capabilities are cached in `device.pyx` so calling this isn't time-consuming. 

Yeah I noticed that too...

> IMHO, this dictionary-based implementation is a little complex but no performance gain. What do you think? Do you have any idea to reduce `device.get_device_id()` calls?

You idea above would work, but I am not sure if it's less complex. I found `device.get_device_id()` is called frequently in a number of places, including in `reduction.pxi`, so perhaps it's not that bad to call it often? (And how often are we considering detrimental anyway?)

If it is the case, then a better approach to achieve what we're looking for is to simply decorate with `memoize()`:
```python
from cupy.util import memoize
@memoize(for_each_device=True)
cdef _cub_reduce_dtype_compatible(x_dtype, int op, dtype=None):
```
Could you verify if this works?

You could also wait for input from the reviewer. They should be more knowledgable. 

One side note: I noticed there's a 4th argument `segmented` in `_cub_reduce_dtype_compatible`. I don't think it's used, and I can't recall why I added it...😛 Could you please also remove it for me if you can confirm it's redundant?

Reply:
> I found device.get_device_id() is called frequently in a number of places, including in reduction.pxi, so perhaps it's not that bad to call it often? (And how often are we considering detrimental anyway?)

Okay, I change my mind to call get_device_id() many times.

I pushed the commit like your dictionary-based idea. And I also removed `segmented` argument.

Can we change `get_device_id()` implementation to cache current device id on python-side?
I think current CuPy implementation calls this function too many times.

Reply:
Thank you, @y1r!

> Can we change `get_device_id()` implementation to cache current device id on python-side?
I think current CuPy implementation calls this function too many times.

I thought about this too. It'd be nice to have this cache, like what is done for streams and cuFFT plans, which can also be used as a context manager like devices. Under the hood the cache is done on a per-thread basis, so some care must be taken. 

Anyway, this change should be made in a separate PR. You can open a new issue to track this problem, and then, if you want, work on a PR that closes the issue. (Just say "This PR closes #XXXX" in the PR description, with XXXX being the issue number. GitHub will ensure the issue is automatically closed when the PR is merged.) Tag me if you create a PR and need help/comment.